### PR TITLE
Brewing Stands

### LINKED ISSUE #520: Implement brewing stands

### PR DESCRIPTION
**Pull request by xNatsuri**
This PR implements brewing stands

Built off #632 
Resolves #520

### CODE REVIEW

File: server/conf.go [hunk 1]
```diff
@@ -157,6 +157,8 @@ func (conf Config) New() *Server {
 	}
 	world_finaliseBlockRegistry()
 
+	// this is not a solution this is just for me to be able to test without having an import cycle.
+	// note: everything works but creative inventory is not showing up properly content wise.
 	recipe.RegisterVanilla()
```
Comment (xNatsuri) [inline server/conf.go:162]: still need to find a solution to this issue

File: server/item/recipe/vanilla.go [hunk 2]
```diff
@@ -45,7 +46,21 @@ type furnaceRecipe struct {
...
+	Reagent inputItem  `nbt:"reagent"`
+	Output  outputItem `nbt:"output"`
+}
+
+// potionContainerChangeRecipe ...
+type potionContainerChangeRecipe struct {
+	Input   string    `nbt:"input"`
+	Reagent inputItem `nbt:"reagent"`
+	Output  string    `nbt:"output"`
+}
```
Comment (TwistedAsylumMC) [inline server/item/recipe/vanilla.go:61]: Could you finish these docs?

File: server/session/player.go [hunk 3]
```diff
@@ -106,6 +106,11 @@ func (s *Session) SendRespawn(pos mgl64.Vec3) {
 // sendRecipes sends the current crafting recipes to the session.
 func (s *Session) sendRecipes() {
 	recipes := make([]protocol.Recipe, 0, len(recipe.Recipes()))
+
+	//temporary solution until I find a better way to handle potion recipes and potion container change recipes.
+	potionRecipes := make([]protocol.PotionRecipe, 0)
+	potionContainerChange := make([]protocol.PotionContainerChangeRecipe, 0)
+
```
Comment (TwistedAsylumMC) [inline server/session/player.go:113]: Is this still temporary or is there something you'd like to change?

**Review by TwistedAsylumMC — CHANGES_REQUESTED** [review]
Just a few things, staticcheck seems to also be kicking up a fuss about some things but not really sure what's up with that

File: server/session/player.go (hunk 3 above)
Comment (xNatsuri) [reply server/session/player.go:113]: it was temporary I'll remove the comment.

**Review by TwistedAsylumMC — APPROVED** [review]

File: server/item/recipe/register.go [hunk 4]
```diff
@@ -15,4 +27,91 @@ func Recipes() []Recipe {
...
+
+	if okTwo {
+		stack := p.Input()[1].(item.Stack)
+		name, _ := stack.Item().EncodeItem()
+		reagent[name] = stack
+	}
+
+	if ok || okTwo {
+		_, containerChange := recipe.(PotionContainerChange)
+		inputItems2 := make([]world.Item, len(recipe.Input()))
```
Comment (DaPigGuy) [inline server/item/recipe/register.go:42]: Why is this named inputItems2?

**Review by DaPigGuy — CHANGES_REQUESTED** [review]

File: server/item/recipe/register.go [hunk 5]
```diff
@@ -15,4 +27,91 @@ func Recipes() []Recipe {
...
+		if useMeta {
+			a := *(*[2]byte)(unsafe.Pointer(&meta))
+			b.Write(a[:])
+		}
+	}
+	return b.String()
+}
+
+// ValidBrewingReagent checks if the world.Item is a brewing reagent.
+func ValidBrewingReagent(itemName string) bool {
```
Comment (DaPigGuy) [inline server/item/recipe/register.go:114]: Why not take a `world.Item` here?

File: server/item/recipe/register.go [hunk 6]
```diff
@@ -15,4 +27,91 @@ func Recipes() []Recipe {
...
+	if !ok {
+		// Block specific index didn't exist.
+		return nil, false
+	}
+	r, ok := blockInd[hashItems(input, true)]
+	if !ok {
+		r, ok = blockInd[hashItems(input, false)]
+		if !ok {
+			return nil, false
+		}
```
Comment (DaPigGuy) [inline server/item/recipe/register.go:71]: This is problematic for potion recipes since potions share the same id

File: server/block/brewer.go [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+			for brewingSlot, brewingStack := range b.inventory.Slots() {
+				if brewingSlot == 0 || brewingSlot == 4 || !brewingStack.Empty() {
+					continue
+				}
+
+				_, okPotion := sourceStack.Item().(item.Potion)
+				_, okSplash := sourceStack.Item().(item.SplashPotion)
+				if !okPotion && !okSplash {
+					continue
+				}
```
Comment (DaPigGuy) [inline server/block/brewer.go:62]: Move this out of the for loop. Also allow lingering potions

Comment (xNatsuri) [reply server/block/brewer.go]: fixed with latest commit

**Review by DaPigGuy — APPROVED** [review]

**Pull request merged**